controllers/actions.github.com/autoscalingrunnerset_controller.go

```diff
@@ -305,6 +318,203 @@ func (r *AutoscalingRunnerSetReconciler) deleteEphemeralRunnerSets(ctx context.C
 	return nil
 }
 
+func (r *AutoscalingRunnerSetReconciler) removeFinalizersFromPermissionResources(ctx context.Context, autoscalingRunnerSet *v1alpha1.AutoscalingRunnerSet, logger logr.Logger) (requeue bool, err error) {
+	requeue, err = r.removeFinalizersContainerModeKubernetesPermissions(ctx, autoscalingRunnerSet, logger)
+	if err != nil {
+		return true, err
+	}
+	if requeue {
+		return true, nil
+	}
+
+	if serviceAccountName, ok := autoscalingRunnerSet.Annotations[AnnotationKeyNoPermissionServiceAccountName]; ok {
+		logger.Info("Removing finalizer from no permission service account")
+		serviceAccount := new(corev1.ServiceAccount)
+		err = r.Get(ctx, types.NamespacedName{Name: serviceAccountName, Namespace: autoscalingRunnerSet.Namespace}, serviceAccount)
+		switch {
+		case err == nil:
+			if !controllerutil.ContainsFinalizer(serviceAccount, autoscalingRunnerSetCleanupFinalizerLabel) {
+				break
+			}
+			err = patch(ctx, r.Client, serviceAccount, func(obj *corev1.ServiceAccount) {
+				controllerutil.RemoveFinalizer(obj, autoscalingRunnerSetCleanupFinalizerLabel)
+			})
+			return true, err
+		case err != nil && !kerrors.IsNotFound(err):
+			return true, err
+		}
+		logger.Info("Removed finalizer from no permission service account", "name", serviceAccountName)
+	} else {
+		logger.Info(
+			"Skipping cleaning up no permission service account",
+			"reason",
+			fmt.Sprintf("annotation key %q not present", AnnotationKeyGitHubSecretName),
+		)
+	}
+
+	if githubSecretName, ok := autoscalingRunnerSet.Annotations[AnnotationKeyGitHubSecretName]; ok {
+		logger.Info("Removing finalizer from GitHub secret")
+
+		secret := new(corev1.Secret)
+		err = r.Get(ctx, types.NamespacedName{Name: githubSecretName, Namespace: autoscalingRunnerSet.Namespace}, secret)
+		switch {
+		case err == nil:
+			if !controllerutil.ContainsFinalizer(secret, autoscalingRunnerSetCleanupFinalizerLabel) {
+				break
+			}
+			err = patch(ctx, r.Client, secret, func(obj *corev1.Secret) {
+				controllerutil.RemoveFinalizer(obj, autoscalingRunnerSetCleanupFinalizerLabel)
+			})
+			return true, err
+		case err != nil && !kerrors.IsNotFound(err) && !kerrors.IsForbidden(err):
+			return true, err
+		}
+		logger.Info("Removed finalizer from GitHub secret", "name", githubSecretName)
+	} else {
+		logger.Info(
+			"Skipping cleaning up GitHub secret",
+			"reason",
+			fmt.Sprintf("annotation key %q not present", AnnotationKeyGitHubSecretName),
+		)
+	}
```

Comment:
it would nice to refactor this into some generic common method since we are using the same pattern to remove finalizers for all resources.